docs/core/extensions/options-library-authors.md

```diff
@@ -0,0 +1,125 @@
+ms.date: 01/20/2021
+---
+
+# Options pattern guidance for .NET library authors
+
+With the help of dependency injection, registering your services and their corresponding configurations can make use of the *options pattern*. The options pattern enables consumers of your library (and your services) to require instances of <xref:Microsoft.Extensions.Options.IOptions%601> where `TOptions` is your options class. As a .NET library author, you'll learn general guidance on how to correctly expose the options pattern to consumers of your library. There are various ways to achieve the same thing, and several considerations to make.
+
+## Naming conventions
+
+By convention, extension methods responsible for registering services are named `Add{Service}`, where `{Service}` is a meaningful and descriptive name. Depending on the package, the registration of services may be accompanied by `Use{Service}` extension methods. The `Use{Service}` extension methods are commonplace in [ASP.NET Core](/aspnet).
+
+✔️ CONSIDER names that disambiguate your service from other offerings.
+
+❌ DO NOT use names that are already part of the .NET ecosystem from official Microsoft packages.
+
+✔️ CONSIDER naming static classes that expose extension methods as `{Type}Extensions`, where `{Type}` is the type that you're extending.
+
+### Namespace guidance
+
+Microsoft packages make use of the `Microsoft.Extensions.DependencyInjection` namespace to unify the registration of various service offerings.
+
+✔️ CONSIDER a namespace that clearly identifies your package offering.
+
+❌ DO NOT use the `Microsoft.Extensions.DependencyInjection` namespace for non-official Microsoft packages.
+
+## Parameterless
+
+If your service can work with minimal or no explicit configuration, consider a parameterless extension method.
+
+:::code language="csharp" source="snippets/configuration/options-noparams/ServiceCollectionExtensions.cs" highlight="10-14":::
+
+In the preceding code, the `AddMyLibraryService`:
+
+- Extends an instance of <xref:Microsoft.Extensions.DependencyInjection.IServiceCollection>
+- Calls <xref:Microsoft.Extensions.DependencyInjection.OptionsServiceCollectionExtensions.AddOptions%60%601(Microsoft.Extensions.DependencyInjection.IServiceCollection)?displayProperty=nameWithType> with the type parameter of `LibraryOptions`
+- Chains a call to <xref:Microsoft.Extensions.Options.OptionsBuilder%601.Configure%2A>, which specifies the default option values
+
+As the library author, specifying default values is up to you.
+
+> [!NOTE]
+> It is possible to bind configuration to an options instance. However, there is a risk of name collisions - which will cause errors. Additionally, when manually binding in this way, you limit the consumption of your options pattern to read-once. Changes to settings will not be re-bound, as such consumers will not be able to use the [IOptionsMonitor](options.md#ioptionsmonitor) interface.
+>
+> ```csharp
+> services.AddOptions<LibraryOptions>()
```

Comment:
I think we should do a better job talking about the various ways users may configure your component once you use Options, and what expectations they might have.

Reply:
How would you phrase this? In what ways might consumers configure the component, are you talking about the various config providers? What expectations would they have, I'm not really sure?

Reply:
Yes the intro section should describe the benefits of using options in your library and explain what options consumers of the library have configuring it. For example, they can configure options without you providing a specific overload for handling configuration.